### ISSUE TITLE
Indentation removed on empty lines when Format Document is used

### ISSUE THREAD

**Issue by poodle-chowder** [OmniSharp/omnisharp-vscode#1980]
![Imgur](https://i.imgur.com/c45Zb9f.gif)

VS Code version: Code 1.19.2 (490ef761b76b3f3b3832eff7a588aac891e5fe80, 2018-01-10T15:55:03.538Z)
C# Extension version: 1.13.1

## Steps to reproduce
Make multiple new lines in an indented code block, then

- right click the workspace and choose "Format Document"
- use FormatOnType, FormatOnPaste, etc., to automatically format the document
- press Ctrl-Shift-I to auto indent document

## Expected  behavior
All lines retain their indentation

## Actual behavior
Any lines in the document without code have their indentation removed

I have narrowed this down to this plugin, as this behavior isn't seen in VSCode when extension is disabled. I have tried every combination of white space trimming, etc., with no solution.

See [Reddit thread](https://www.reddit.com/r/vscode/comments/7sifqk/indentation_problems/).

**Comment by akshita31** [comment]
Possible duplicate of https://github.com/OmniSharp/omnisharp-vscode/issues/1680

**Comment by poodle-chowder** [comment]
See text highlighting at end of attached gif, doesn't appear to be caret positioning but actual deletion of the spaces.

**Comment by DustinCampbell** [comment]
@poodle-chowder : The goal of the whitespace formatter is to remove superfluous whitespace. So, the behavior that you're seeing where unnecessary indentation at the starts of lines is removed is working as designed. I suspect that the real issue here is really the same as #1680, where the indentation is removed on the line where the caret appears, shifting it to the left.

**Comment by poodle-chowder** [comment]
@DustinCampbell Understood. I think the heart of the issue is that whether `omnisharp.useEditorFormattingSettings` is turned on or off, editor whitespace trimming settings are ignored. With the omnisharp extension disabled, these settings are honored and I don't have the indentation issue.

VS does not seem to consider the indents leading up to a block's indentation level to be superfluous, but VSCode formats to 0 indent even in the middle of a code block. This seems like odd behavior to me. Is the agenda to keep workflow consistent between the two?

**Comment by DustinCampbell** [comment]
Not exactly. The intent is to keep the formatting similar _for C#_ between Visual Studio and Visual Studio Code. C#'s formatter has always removed extra indentation, but there's special behavior on the line that the caret is on.

**Comment by fizzd** [comment]
Is there any alternative to format documents without removing whitespace on empty lines now? Or even just a 'format without removing whitespace anywhere at all'? I have tried a lot of solutions and disabling/enabling whitespace trimming settings, but whitespace on blank lines is always deleted.

**Comment by SonOfHans** [comment]
I would like to have the option to _not_ have whitespace removed on empty lines, regardless of where the caret is located. I frequently leave empty lines between code blocks so as to make my code more readable, oftentimes adding additional code in between these blocks at a later time. Having to re-tab a bunch of times to start that new block is definitely not preferred. This was a feature in VS, and I miss it greatly in VS Code.

**Comment by eric-vgastel** [comment]
I've been looking for a solution for weeks. It's driving me literally insane. Please fix this.

**Comment by radutomy** [comment]
I, too, would like a solution to this. It's very, very annoying, especially when combined with Vim.

**Comment by davidmilligan** [comment]
This bug is infuriating, and nothing has been done about it for years?! The people who maintain this project must not actually use it for real-world, day-to-day work. I don’t see how anyone can stand this behavior. It seems like this issue has just been blamed on some other issue and so nothing is being done about it. Unacceptable people!

**Comment by tr0q** [comment]
I just checked Visual Studio, indeed it deletes whitespaces, it's just it keeps caret on correct indentation level.  So it's indeed not an extension fault but VSCode itself.

**Comment by wbplomp** [comment]
I like empty lines to be removed (with format document) while editing JSON files. Now for some reason this has behavior has been removed.

**Comment by geekley** [comment]
This wouldn't be an issue if VSCode had a behavior similar to IntelliJ/Android Studio.  
When you go on a blank line (not indented) and press END, it "pretends" to auto-add the whitespace for the correct indentation, and after you type something, it actually adds it for you. So, even though the whitespace is trimmed on blank lines, it's not an annoyance.

**Comment by Creativogee** [comment]
It's 2020 and there is still no fix to this bug as i am currently experiencing the same issue. I can't believe it. This is literally driving me nuts. One moment it was working the next I know it wasn't.

**Comment by nzeman** [comment]
Any news on this? I cannot switch to VSCode with this. It would drive me insane.

**Comment by Phyyl** [comment]
I also would like this to be fixed. I really think having "fake" indentation is the way to go.

**Comment by JoeRobich** [comment]
I encourage everyone interested in this fuctionality to upvote the following vscode issue - https://github.com/microsoft/vscode/issues/13960

**Comment by paweston70** [comment]
> I encourage everyone interested in this fuctionality to upvote the following vscode issue - [microsoft/vscode#13960](https://github.com/microsoft/vscode/issues/13960)

But virtual space is not the same thing - this issue highlights that VS Code is formatting out indents on blank lines...

**Comment by Hezkore** [comment]
I thought this was specific to the formatting tool used, but if not I certainly with this was an option.
Don't mess with my indentation!

**Comment by cornytrace** [comment]
This is a very annoying problem, and makes interoperability with Visual Studio almost impossible without changing whitespace back and forth in each commit.

**Comment by BachiMjavanadze** [comment]
Please fix it

**Comment by RandomGgames** [comment]
Please fix!

**Comment by IXLLEGACYIXL** [comment]
Still doesnt work

**Comment by acdupont** [comment]
> The goal of the whitespace formatter is to remove superfluous whitespace

Is it not arguable that there are reasons to consider indents on empty lines as non superfluous? If it is arguable at all, then there should be a configuration option for it. Forcing one view to be the solution for everyone in all contexts is extremely frustrating.

**Comment by bernd4** [comment]
Maybe if we use some fucking rough language this fucking shit issue will get some fucking attention!
It's fucking still fucking shitty anoying!
FUCK!

**Comment by VMaldecoasago** [comment]
Any chance of prioritizing this? It's a very annoying issue, indeed.

**Comment by victoraldecoa** [comment]
Any chance of prioritizing this? It's a very annoying issue, indeed.

**Comment by mana-break** [comment]
When I click on a blank line in Visual Studio (Mac) the caret automatically gets a virtual indent from the previous line. In vscode the caret doesn't get any indentation and I have to press tab several times. It infuriates me. Make it behave like in Visual Studio

**Comment by geekley** [comment]
> Is it not arguable that there are reasons to consider indents on empty lines as non superfluous?

Yes, it's awful. When I leave blank space between methods in a class, I want to keep it at that class' indentation, so that it's already right when inserting new code. If you think this space is unnecessary, I disagree, and in fact I think *removing it* is what's not only unnecessary but also counter-intuitive - unless the user explicitly wants to do it this way.

**Comment by larrysammii** [comment]
It's been 6 years...
I guess I'm just gonna switch editor.

**Comment by geewasvoboda** [comment]
I would also like to see this feature, the auto formatting is basically unusable because of this

**Comment by jr-uk** [comment]
Finally found a decent simple way this great extension: https://marketplace.visualstudio.com/items?itemName=jemc.vscode-implicit-indent

**Comment by StoneWolf** [comment]
> Finally found a decent simple way with this great extension: https://marketplace.visualstudio.com/items?itemName=jemc.vscode-implicit-indent

Be warned that this extension breaks your undo/redo history. If you undo a bunch of edits then move your caret to an empty line, you'll lose all your redo history.

**Comment by claybford** [comment]
Any chance this can be prioritized? Bueller?

**Comment by JoeRobich** [comment]
This is the VSCode issue tracking Smart Indent. If the feature is meaningful to you, then be sure to let them know. https://github.com/microsoft/vscode/issues/16998

**Comment by jonathandlo** [comment]
hi, please add an option to retain whitespace on blank lines, or have the formatter configurable to add whitespace to match the expected indent level

**Comment by aeronhibby** [comment]
fix this please, I hate it.

**Comment by ghost** [comment]
Just came across this problem due to formatting issues. I have no problem with the indentation being removed from a file but the issue is that when pressing `Enter` to go to the next line to write code, the indentation does not update to the correct level. I have to press `Tab` a few times to match the indentation level. Kind of frustrating considering Rider does this automatically.

**Comment by aeronhibby** [comment]
Installing [this](https://marketplace.visualstudio.com/items?itemName=LAK132.indent-to-bracket) in extensions fixed it for me

**Comment by NorthboundGames** [comment]
> Installing [this](https://marketplace.visualstudio.com/items?itemName=LAK132.indent-to-bracket) in extensions fixed it for me

This changes nothing. The C# formatter already indents to the correct position after a carriage return. The issue is that format document deletes whitespace so when you click to insert code on an empty line you have to manually indent. It might sound trivial, but try doing it hundreds of times a day. Every other editor I've used in the 7 years since I posted this issue will honor whitespace on empty lines.

**Comment by geekley** [comment]
Ah, seems like I forgot about this and opened another issue at the roslyn repo. dotnet/roslyn#83443

I'll consider it not a duplicate because in there I'm proposing a .editorconfig setting for the formatter:
`csharp_indent_blank_lines = none|ignore|indent`
I'm also arguing that on blank lines, the indentation whitespace is not "trailing whitespace", it's "leading whitespace"; only extra whitespace AFTER the indentation is "trailing" to be removed by a formatter.

So my proposal there is not for the formatter to just leave it alone necessarily, rather you should be able to choose between:
- `indent`: formatter should fix all blank indentation to where code would be if it wasn't blank
- `none`: current behavior, in case you want to minimize a couple of file bytes (can't see any other reason to use this)
- `ignore`: just leave my spaces alone! (on blank lines) don't fix it one way or the other

 Commenting here in case y'all want to upvote it.

**Comment by acdupont** [comment]
> I'm also arguing that on blank lines, the indentation whitespace is not "trailing whitespace", it's "leading whitespace"; only extra whitespace AFTER the indentation is "trailing" to be removed by a formatter.

Exactly! There is no reason the developer shouldn't control how they want this handled.
